src/common/classes/BatchCompletionState.h

```diff
@@ -0,0 +1,185 @@
+#include "../common/classes/array.h"
+#include "../common/utils_proto.h"
+
+namespace Firebird {
+
+	class Transliterate
+	{
+	public:
+		virtual void transliterate(IStatus* status) = 0;
+	};
+
+	class BatchCompletionState FB_FINAL :
+    	public DisposeIface<Firebird::IBatchCompletionStateImpl<BatchCompletionState, CheckStatusWrapper> >
+	{
+	public:
+		BatchCompletionState(bool storeCounts, ULONG lim)
+			: rare(getPool()),
+			  reccount(0u),
+			  detailedLimit(lim)
+		{
+			if (storeCounts)
+				array = FB_NEW_POOL(getPool()) DenseArray(getPool());
+		}
+
+		void regError(IStatus* errStatus, Transliterate* transliterate)
+		{
+			IStatus* newVector = nullptr;
+			if (rare.getCount() < detailedLimit)
+			{
+				newVector = errStatus->clone();
+				if (transliterate)
+					transliterate->transliterate(newVector);
+			}
+			rare.add(StatusPair(reccount, newVector));
+
+			regUpdate(IBatchCompletionState::EXECUTE_FAILED);
+		}
+
+		void regErrorAt(ULONG at, IStatus* errStatus)
+		{
+			IStatus* newVector = nullptr;
+			if ((rare.getCount() < detailedLimit) && errStatus)
+				newVector = errStatus->clone();
+			rare.add(StatusPair(at, newVector));
+		}
+
+		void regUpdate(SLONG count)
+		{
+			if (array)
+				array->push(count);
+
+			++reccount;
+		}
+
+		void regSize(ULONG total)
+		{
+			reccount = total;
+		}
+
+		// IBatchCompletionState implementation
+		unsigned getSize(CheckStatusWrapper*)
+		{
+			return reccount;
+		}
+
+		int getState(CheckStatusWrapper* status, unsigned pos)
+		{
+			try
+			{
+				if (pos >= reccount)
+					(Arg::Gds(isc_random) << "Position is out of range").raise();
+				if (array)
+					return (*array)[pos];
+
+				ULONG index = find(pos);
+				return (index >= rare.getCount() || rare[index].first != pos) ?
+					SUCCESS_NO_INFO : EXECUTE_FAILED;
+			}
+			catch (const Exception& ex)
+			{
+				ex.stuffException(status);
+			}
+			return 0;
+		}
+
+		unsigned findError(CheckStatusWrapper* status, unsigned pos)
+		{
+			try
+			{
+				ULONG index = find(pos);
+				if (index < rare.getCount())
+					return rare[index].first;
+			}
+			catch (const Exception& ex)
+			{
+				ex.stuffException(status);
+			}
+			return NO_MORE_ERRORS;
+		}
+
+		void getStatus(CheckStatusWrapper* status, IStatus* to, unsigned pos)
+		{
+			try
+			{
+				if (pos >= reccount)
+					(Arg::Gds(isc_random) << "Position is out of range").raise();
+
+				ULONG index = find(pos);
+				if (index < rare.getCount() && rare[index].first == pos)
+				{
+					if (rare[index].second)
+					{
+						CheckStatusWrapper w(to);
+						fb_utils::copyStatus(&w, rare[index].second);
+						return;
+					}
+					(Arg::Gds(isc_random) << "Detailed error info is missing in batch").raise();
+				}
+			}
+			catch (const Exception& ex)
+			{
+				ex.stuffException(status);
+			}
+		}
+
+		void dispose()
+		{
+			delete this;
+		}
+
+		~BatchCompletionState()
+		{
+			for (unsigned i = 0; i < rare.getCount() && rare[i].second; ++i)
```

Comment:
No, it can not. In rare array only first N (default 64) status vectors are kept - rest of the elements always have a NULL second. This is done to avoid server OOM state in case when crazy user has error in all messages.